File: est-modules/est-kotlin-support/est-typescript-sdk/tests/benchmark.test.ts
import { EstClient, PluginMarketplaceClient } from '../src';
import { PluginInfo, PluginReview, SearchOptions, PublishResult, EstClientConfig } from '../src/types';

describe('Benchmark Tests', () => {
  
  test('Client Initialization Benchmark', () => {
    const iterations = 10000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const config: EstClientConfig = { baseUrl: 'http://localhost:8080' };
      const client = new EstClient(config);
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nClient Initialization Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per init`);
    
    expect(avgTime).toBeLessThan(0.1);
  });
  
  test('Client Initialization with API Key Benchmark', () => {
    const iterations = 10000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const config: EstClientConfig = { 
        baseUrl: 'http://localhost:8080',
        apiKey: 'test-api-key'
      };
      const client = new EstClient(config);
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nClient Initialization with API Key Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per init`);
    
    expect(avgTime).toBeLessThan(0.1);
  });
  
  test('PluginMarketplaceClient Initialization Benchmark', () => {
    const iterations = 5000;
    const config: EstClientConfig = { baseUrl: 'http://localhost:8080' };
    const client = new EstClient(config);
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const marketplace = new PluginMarketplaceClient(client);
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nPluginMarketplaceClient Initialization Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per init`);
    
    expect(avgTime).toBeLessThan(0.05);
  });
  
  test('PluginInfo Creation Benchmark', () => {
    const iterations = 100000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const plugin: PluginInfo = {
        name: `Test Plugin ${i}`,
        version: '1.0.0',
        description: 'A test plugin',
        author: 'Test Author',
        mainClass: 'com.example.TestPlugin',
        dependencies: [],
        softDependencies: [],
        category: 'test',
        tags: ['test', 'example'],
        license: 'MIT',
        rating: 4.5,
        downloadCount: 1000,
        screenshots: [],
        publishTime: Date.now(),
        lastUpdateTime: Date.now(),
        certified: true,
        compatibleVersions: ['2.3.0', '2.4.0'],
        minFrameworkVersion: '2.3.0'
      };
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nPluginInfo Creation Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per creation`);
    
    expect(avgTime).toBeLessThan(0.01);
  });
  
  test('PluginReview Creation Benchmark', () => {
    const iterations = 100000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const review: PluginReview = {
        id: `review-${i}`,
        pluginId: 'plugin-1',
        userId: 'user-1',
        userName: 'Test User',
        rating: 5,
        title: 'Great plugin!',
        content: 'Excellent plugin, works perfectly.',
        createdAt: Date.now(),
        updatedAt: Date.now(),
        helpfulCount: 10
      };
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nPluginReview Creation Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per creation`);
    
    expect(avgTime).toBeLessThan(0.01);
  });
  
  test('SearchOptions Creation Benchmark', () => {
    const iterations = 100000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const options: SearchOptions = {
        query: 'web framework',
        category: 'web',
        tags: ['ui', 'framework'],
        page: 2,
        pageSize: 10,
        sortBy: 'rating',
        sortOrder: 'desc'
      };
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nSearchOptions Creation Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per creation`);
    
    expect(avgTime).toBeLessThan(0.005);
  });
  
  test('PublishResult Creation Benchmark', () => {
    const iterations = 100000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const result: PublishResult = {
        success: true,
        pluginId: 'new-plugin',
        version: '1.0.0',
        message: 'Plugin published successfully',
        errors: ['error-1', 'error-2']
      };
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nPublishResult Creation Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per creation`);
    
    expect(avgTime).toBeLessThan(0.005);
  });
  
  test('EstClientConfig Creation Benchmark', () => {
    const iterations = 100000;
    const start = performance.now();
    
    for (let i = 0; i < iterations; i++) {
      const config: EstClientConfig = {
        baseUrl: 'http://localhost:8080',
        apiKey: 'test-api-key',
        timeout: 60000,
        headers: {
          'X-Custom': 'value',
          'Accept': 'application/json'
        }
      };
    }
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nEstClientConfig Creation Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per creation`);
    
    expect(avgTime).toBeLessThan(0.005);
  });
  
  test('Memory Usage Benchmark', () => {
    const iterations = 10000;
    const clients: EstClient[] = [];
    
    const startMemory = process.memoryUsage().heapUsed;
    
    for (let i = 0; i < iterations; i++) {
      const config: EstClientConfig = { baseUrl: 'http://localhost:8080' };
      clients.push(new EstClient(config));
    }
    
    const endMemory = process.memoryUsage().heapUsed;
    const memoryUsed = endMemory - startMemory;
    const avgMemory = memoryUsed / iterations;
    
    console.log('\nMemory Usage Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Total memory: ${(memoryUsed / 1024).toFixed(2)}KB`);
    console.log(`  Average: ${avgMemory.toFixed(2)} bytes per client`);
    
    expect(avgMemory).toBeLessThan(1024);
  });
  
  test('Concurrent Operations Benchmark', async () => {
    const iterations = 100;
    const concurrentTasks = 10;
    
    const createClient = async () => {
      const config: EstClientConfig = { baseUrl: 'http://localhost:8080' };
      return new EstClient(config);
    };
    
    const start = performance.now();
    
    const chunks = [];
    for (let i = 0; i < iterations; i += concurrentTasks) {
      const chunk = [];
      for (let j = 0; j < concurrentTasks && i + j < iterations; j++) {
        chunk.push(createClient());
      }
      chunks.push(Promise.all(chunk));
    }
    
    await Promise.all(chunks);
    
    const elapsed = performance.now() - start;
    const avgTime = elapsed / iterations;
    
    console.log('\nConcurrent Operations Benchmark:');
    console.log(`  Iterations: ${iterations}`);
    console.log(`  Concurrent tasks: ${concurrentTasks}`);
    console.log(`  Total time: ${elapsed.toFixed(2)}ms`);
    console.log(`  Average: ${avgTime.toFixed(4)}ms per operation`);
    
    expect(avgTime).toBeLessThan(1.0);
  });
});

File: est-modules/est-kotlin-support/est-python-sdk/tests/benchmark_test.py
"""
EST Framework Python SDK - 性能基准测试
"""

import pytest
import time
from est_sdk import EstClient
from est_sdk.types import PluginMetadata, PluginSearchCriteria
from unittest.mock import Mock, patch


class TestBenchmark:
    
    @pytest.fixture
    def client(self):
        return EstClient(base_url="http://localhost:8080")
    
    def test_client_initialization_benchmark(self, client):
        iterations = 1000
        start_time = time.perf_counter()
        
        for _ in range(iterations):
            _ = EstClient(base_url="http://localhost:8080")
        
        elapsed = time.perf_counter() - start_time
        avg_time = (elapsed / iterations) * 1000
        
        print(f"\nClient initialization benchmark:")
        print(f"  Iterations: {iterations}")
        print(f"  Total time: {elapsed:.4f}s")
        print(f"  Average: {avg_time:.4f}ms per init")
        
        assert avg_time < 1.0, f"Initialization too slow: {avg_time:.4f}ms"
    
    def test_plugin_metadata_creation_benchmark(self):
        iterations = 10000
        start_time = time.perf_counter()
        
        for i in range(iterations):
            _ = PluginMetadata(
                plugin_id=f"plugin-{i}",
                name=f"Test Plugin {i}",
                description=f"A test plugin {i}",
                version="1.0.0",
                author="Test Author",
                category="test"
            )
        
        elapsed = time.perf_counter() - start_time
        avg_time = (elapsed / iterations) * 1000
        
        print(f"\nPluginMetadata creation benchmark:")
        print(f"  Iterations: {iterations}")
        print(f"  Total time: {elapsed:.4f}s")
        print(f"  Average: {avg_time:.4f}ms per creation")
        
        assert avg_time < 0.1, f"PluginMetadata creation too slow: {avg_time:.4f}ms"
    
    def test_search_criteria_creation_benchmark(self):
        iterations = 10000
        start_time = time.perf_counter()
        
        for i in range(iterations):
            _ = PluginSearchCriteria(
                query=f"query-{i}",
                category="test",
                page=1,
                page_size=20
            )
        
        elapsed = time.perf_counter() - start_time
        avg_time = (elapsed / iterations) * 1000
        
        print(f"\nPluginSearchCriteria creation benchmark:")
        print(f"  Iterations: {iterations}")
        print(f"  Total time: {elapsed:.4f}s")
        print(f"  Average: {avg_time:.4f}ms per creation")
        
        assert avg_time < 0.1, f"Search criteria creation too slow: {avg_time:.4f}ms"
    
    @patch('requests.Session.request')
    def test_request_performance_benchmark(self, mock_request, client):
        mock_response = Mock()
        mock_response.raise_for_status.return_value = None
        mock_response.json.return_value = {"key": "value"}
        mock_request.return_value = mock_response
        
        iterations = 1000
        start_time = time.perf_counter()
        
        for _ in range(iterations):
            _ = client.get("/test")
        
        elapsed = time.perf_counter() - start_time
        avg_time = (elapsed / iterations) * 1000
        
        print(f"\nRequest performance benchmark (mocked):")
        print(f"  Iterations: {iterations}")
        print(f"  Total time: {elapsed:.4f}s")
        print(f"  Average: {avg_time:.4f}ms per request")
        
        assert avg_time < 0.5, f"Request handling too slow: {avg_time:.4f}ms"
    
    def test_concurrent_client_operations(self):
        from concurrent.futures import ThreadPoolExecutor, as_completed
        
        def create_client():
            return EstClient(base_url="http://localhost:8080")
        
        iterations = 100
        start_time = time.perf_counter()
        
        with ThreadPoolExecutor(max_workers=10) as executor:
            futures = [executor.submit(create_client) for _ in range(iterations)]
            for future in as_completed(futures):
                _ = future.result()
        
        elapsed = time.perf_counter() - start_time
        avg_time = (elapsed / iterations) * 1000
        
        print(f"\nConcurrent client operations benchmark:")
        print(f"  Iterations: {iterations}")
        print(f"  Total time: {elapsed:.4f}s")
        print(f"  Average: {avg_time:.4f}ms per operation")
        
        assert avg_time < 5.0, f"Concurrent operations too slow: {avg_time:.4f}ms"
    
    def test_memory_usage_benchmark(self):
        import sys
        
        clients = []
        iterations = 1000
        
        start_memory = sys.getsizeof(clients)
        
        for i in range(iterations):
            client = EstClient(base_url="http://localhost:8080")
            clients.append(client)
        
        end_memory = sys.getsizeof(clients)
        for client in clients:
            end_memory += sys.getsizeof(client)
            if hasattr(client, 'session'):
                end_memory += sys.getsizeof(client.session)
        
        avg_memory = (end_memory - start_memory) / iterations
        
        print(f"\nMemory usage benchmark:")
        print(f"  Iterations: {iterations}")
        print(f"  Total memory: {end_memory / 1024:.2f}KB")
        print(f"  Average: {avg_memory:.2f} bytes per client")
        
        assert avg_memory < 1024, f"Memory usage too high: {avg_memory:.2f} bytes"
